Content.Server/Weapons/Melee/GasPoweredMeleeThrowSystem.cs

```diff
@@ -0,0 +1,82 @@
+using Content.Server.Atmos.Components;
+using Content.Server.Atmos.EntitySystems;
+using Content.Shared.Containers.ItemSlots;
+using Robust.Shared.Containers;
+using Content.Shared.Weapons.Melee.Components;
+using Content.Shared.Weapons.Melee;
+using Robust.Shared.Audio;
+
+namespace Content.Server.Weapons.Melee;
+
+/// <summary>
+/// Checks if the used entity has a gas tank equipped and removes gas from tank when melee throw occurs.
+/// </summary>
+public sealed class GasPoweredMeleeThrowSystem : SharedGasPoweredMeleeThrowSystem
+{
+    [Dependency] private readonly ItemSlotsSystem _itemSlots = default!;
+    [Dependency] private readonly AtmosphereSystem _atmos = default!;
+    [Dependency] private readonly GasTankSystem _gasTank = default!;
+
+    public override void Initialize()
+    {
+        SubscribeLocalEvent<GasPoweredThrowerComponent, AttemptMeleeThrowOnHitEvent>(OnAttemptMeleeThrowOnHit);
+        SubscribeLocalEvent<GasPoweredThrowerComponent, ContainerIsInsertingAttemptEvent>(OnContainerInserting);
+    }
+
+    private void OnContainerInserting(EntityUid uid, GasPoweredThrowerComponent component, ContainerIsInsertingAttemptEvent args)
+    {
+        if (args.Container.ID != component.TankSlotId)
+            return;
+
+        if (!TryComp<GasTankComponent>(args.EntityUid, out var gas))
+            return;
+
+        if (gas.Air.TotalMoles >= component.GasUsage && component.GasUsage > 0f)
+            return;
+
+        args.Cancel();
+    }
```

Comment:
Silently mispredicting due to serverside code is really awful and should be providing feedback.